### ISSUE TITLE
Flutter macOS template depends on a private NSViewController initializer

### ISSUE THREAD

**Issue by rundong08** [flutter/flutter#96128]
In the file https://github.com/flutter/flutter/blob/73fb2e653dd058e97dfc81954aa944a15c2dae5b/packages/flutter_tools/templates/app_shared/macos.tmpl/Runner/MainFlutterWindow.swift#L6

The initializer `init()` is not defined for class `FlutterViewController`, see [FlutterViewController.mm](https://github.com/flutter/engine/blob/fa558f6a7ea694caff289d9e589b576a8515ea9b/shell/platform/darwin/macos/framework/Source/FlutterViewController.mm) and [FlutterViewController.h](https://github.com/flutter/engine/blob/fa558f6a7ea694caff289d9e589b576a8515ea9b/shell/platform/darwin/macos/framework/Headers/FlutterViewController.h). The `init()` initializer is also not publicly defined for `FlutterViewController`'s parent class `NSViewController` (see https://developer.apple.com/documentation/appkit/nsviewcontroller and the `NSViewController.h` file in the SDK). The `init()` initializer is defined for `NSViewController`'s parent class `NSResponder`. But that initializer certainly will not initialize a `FlutterViewController` object properly. 

By some step-through debugging, it turns out  that `FlutterViewController.init()` is actually dispatched to `NSViewController.init()` method, which is neither documented nor appearing in the header file. And that `init` method happened to call the instance's `initWithNibName:bundle:` initializer (with nil parameter values), which happened to be implemented by  `FlutterViewController`.

If Apple changes the existence or implementation of the `NSViewController.init()` method, this`init()` might fail. Changing it to an explicit call to the `initWithNibName:bundle:` initializer would solve the issue.

**Comment by jmagman** [comment]
> But that initializer certainly will not initialize a FlutterViewController object properly.

What issue exactly are you seeing?  `init()` is the `NSResponder` designated initializer, which makes it also a designated initializer for `NSViewController` (it's not redeclared as a convenience initializer, nor marked unavailable).

**Comment by rundong08** [comment]
> What issue exactly are you seeing? `init()` is the `NSResponder` designated initializer, which makes it also a designated initializer for `NSViewController` (it's not redeclared as a convenience initializer, nor marked unavailable).

`NSResponder`'s `init()` method will not initialize `FlutterViewController`'s properties such as `_project` and `_engine`.

The actual initializer called is not from `NSResponder`, but from `NSViewController` itself, defined as (dumped from Xcode's debugger)
```asm
AppKit`-[NSViewController init]:
->  0x7fff22ebdae7 <+0>:  pushq  %rbp
    0x7fff22ebdae8 <+1>:  movq   %rsp, %rbp
    0x7fff22ebdaeb <+4>:  movq   0x5dd8b5be(%rip), %rsi    ; "initWithNibName:bundle:"
    0x7fff22ebdaf2 <+11>: xorl   %edx, %edx
    0x7fff22ebdaf4 <+13>: xorl   %ecx, %ecx
    0x7fff22ebdaf6 <+15>: popq   %rbp
    0x7fff22ebdaf7 <+16>: jmpq   *0x6459a083(%rip)         ; (void *)0x00007fff20400800: objc_msgSend
```
which calls the instance's `initWithNibName:bundle:` method, which happened to be implemented by `FlutterViewController`. However, the existence of `[NSViewController init]` is not documented and its implementation that calls the instance's `initWithNibName:bundle:` method is also not guaranteed.

**Comment by stuartmorgan** [comment]
> The `init()` initializer is also not publicly defined for `FlutterViewController`'s parent class `NSViewController`

Yes it is; it's declared in `NSObject`. Objective-C convention is not to redeclare inherited initializers in subclass headers. The assertion made in the title of this issue not correct.

> But that initializer certainly will not initialize a `FlutterViewController` object properly.

I'm not sure how you've reached that conclusion given that we override the designated initializers.

> [...] is actually dispatched to `NSViewController.init()` method, which is neither documented nor appearing in the header file.

There is no requirement to document or redeclare inherited methods in Objective-C, and it is not standard practice.

> And that `init` method happened to call the instance's `initWithNibName:bundle:` initializer (with nil parameter values)

It does not "happen to call" that method. It calls that method because that's a designated initializer for `NSViewController`.

> which happened to be implemented by `FlutterViewController`.

It does not "happen to be implemented". `FlutterViewController overrides it because it's a designated initializer of the superclass.

> If Apple changes the existence or implementation of the `NSViewController.init()` method, this`init()` might fail.

If Apple changed `NSViewController` such that it didn't use designated initializers correctly, it would be violating the API contract of `NSViewController`.

> Changing it to an explicit call to the `initWithNibName:bundle:` initializer would solve the issue.

As far as I can tell you haven't described an actual issue; you've made several assertions that suggest a misunderstanding of how method inheritance and designated initializers work.

**Comment by rundong08** [comment]
> If Apple changed NSViewController such that it didn't use designated initializers correctly, it would be violating the API contract of NSViewController.

Since `NSViewController.init()` is not publicly documented, Apple can do anything with it, without violating the contract about designated initializers.  For example it can be defined as private designated initializer, not calling `initWithNibName:bundle:`.

Please see the following proof of concept:
```objc
#import <Foundation/Foundation.h>

// MockNSResponder.h
@interface MockNSResponder : NSObject
- (instancetype)init NS_DESIGNATED_INITIALIZER;
@end
//------------------

// MockNSResponder.m (private)
@implementation MockNSResponder
- (instancetype)init
{
  NSLog(@"MockNSResponder.init() was called");
  return self;
}
@end
//------------------

// MockNSViewController.h
@interface MockNSViewController : MockNSResponder
- (instancetype)initWithNibName:(nullable NSString*)nibNameOrNil bundle:(nullable NSString *)nibBundleOrNil NS_DESIGNATED_INITIALIZER;
@end
//-----------------------

// MockNSViewController.m (private)
@interface MockNSViewController()
- (instancetype)init NS_DESIGNATED_INITIALIZER;
@end

@implementation MockNSViewController
- (instancetype)initWithNibName:(nullable NSString*)nibNameOrNil bundle:(nullable NSString *)nibBundleOrNil
{
  NSLog(@"MockNSViewController.init(NibName: %@, bundle: %@) was called.", nibNameOrNil, nibBundleOrNil);
  return [super init];
}

-(instancetype) init;
{
  NSLog(@"MockNSViewController.init() was called, but I am not calling the \"initWithNibName:bundle:\" method");
  return [super init];
}
@end
//-----------------------


// MockFlutterViewController.h
@interface MockFlutterViewController : MockNSViewController
- (instancetype)initWithNibName:(nullable NSString*)nibNameOrNil bundle:(nullable NSString *)nibBundleOrNil NS_DESIGNATED_INITIALIZER;
@end
//----------------------------

// MockFlutterViewController.m
@implementation MockFlutterViewController
- (instancetype)initWithNibName:(nullable NSString*)nibNameOrNil bundle:(nullable NSString *)nibBundleOrNil
{
  NSLog(@"MockFlutterViewController.init(NibName: %@, bundle: %@) was called.", nibNameOrNil, nibBundleOrNil);
  return [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
}
@end
//----------------------------


int main(int argc, const char * argv[]) {
  @autoreleasepool {
    MockFlutterViewController* controller = [[MockFlutterViewController alloc] init];
  }
  return 0;
}
```

And its running result is
```
2022-01-04 14:22:34.053933-0800 test-init[65776:1639031] MockNSViewController.init() was called, but I am not calling the "initWithNibName:bundle:" method
2022-01-04 14:22:34.055278-0800 test-init[65776:1639031] MockNSResponder.init() was called
```

**Comment by stuartmorgan** [comment]
> Since `NSViewController.init()` is not publicly documented (explicitly)

I don't understand what distinction you are trying to draw here by saying "explicitly". `init` is part of the public API because that's how ObjC works. Whether it's in the header, or has new documentation, is completely irrelevant here.

> Apple can do anything with it, without violating the contract about designated initializers.

This statement is not correct.

> For example it can be defined as private designated initializer, not calling `initWithNibName:bundle:`.

Designated initializers annotations are necessary for subclassers; making them private makes no sense. What you are describing is Apple intentionally making it impossible to subclass `NSViewController`.

I agree that if that they did that we would be broken, but that's not a realistic case. If Apple decided for some reason to actively break subclasses that could do so in lots of different ways. They could check types in initializers and crash if it was a subclass, for instance.

**Comment by rundong08** [comment]
Thanks for your patient reply, and now I get the point.

When I first looked at the call to `FlutterViewController.init()` it really confused me.  I tried to look up its definition in Xcode, but it all went up to `NSResponder`'s `init()` function. Now I figured out that the correct thought process is: since init() is a designated initializer in NSObject and also NSResponder, its child class (NSViewController) should either override it or making it convenience or unavailable. But since Apple didn't mention it in the document, it must be a convenience initializer, because otherwise the child class of NSViewController won't know about the private designated initializer and cannot inherit correctly.  And since NSViewController.init() is a convenience initializer, it must call a designated initializer. NSViewController has two designated initializer initWithNibName:bundle: and initWithCoder. But apparently initWithNibName:bundle:  is the right choice. So NSViewController.init() must call initWithNibName:bundle:, and it has to pass nil to both parameters since there are no other obvious values.  It makes sense but I feel there are too many assumptions and indirectness there. But that might be just because I am not an expert in Objective-C and Swift.
